substrate/client/authority-discovery/src/worker/addr_cache.rs

```diff
@@ -45,6 +45,66 @@ pub(crate) struct AddrCache {
 	peer_id_to_authority_ids: HashMap<PeerId, HashSet<AuthorityId>>,
 }
 
+impl Serialize for AddrCache {
+	/// We perform SerializeAddrCache::from(self) and then
+	/// we serialize SerializeAddrCache which derives Serialize
+	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+	where
+		S: serde::Serializer,
+	{
+		SerializeAddrCache::from(self.clone()).serialize(serializer)
+	}
+}
+
+impl<'de> Deserialize<'de> for AddrCache {
+	/// We deserialize a `SerializeAddrCache` and then
+	/// we reconstruct the original `AddrCache`.
+	fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+	where
+		D: serde::Deserializer<'de>,
+	{
+		SerializeAddrCache::deserialize(deserializer).map(Into::into)
+	}
+}
+
+/// A private helper struct which is a storage optimized variant of `AddrCache`
+/// which contains the bare minimum info to reconstruct the AddrCache. We rely
+/// on the fact that the `peer_id_to_authority_ids` can be reconstructed from
+/// the `authority_id_to_addresses` field.
+#[derive(Serialize, Deserialize)]
+struct SerializeAddrCache {
```

Comment:
If you prefer I can remove this helper type, and do put the logic directly into `impl Serialize/Deserialize for AddrCache` - however I think it is less clear. But maybe that is just my personal coding preference. 

WDYT?

Reply:
And when using much smaller cache - perhaps close to production like scenario the storage optimized is stil:


```rust
let n: u64 = 300; // authority ID count
let m: u64 = 30; // Multiaddr/Authority count
```

which yields same good results:

```rust
SERIALIZE: Naive took: 16
SERIALIZE: Storage optimized took: 3
DESERIALIZE: Naive took: 10
DESERIALIZE: Storage optimized took: 5
```

Reply:
And for very large cache (so large it is irrelevant to the use case):

```rust
let n: u64 = 1_000_000;
let m: u64 = 20;
```

```
SERIALIZE: Naive took: 37 s
SERIALIZE: Storage optimized took: 10 s
DESERIALIZE: Naive took: 26 s
DESERIALIZE: Storage optimized took: 14 s
```